consensus/keccak/sealer.go

```diff
@@ -0,0 +1,372 @@
+		if err != nil {
+			keccak.lock.Unlock()
+			return err
+		}
+		keccak.rand = rand.New(rand.NewSource(seed.Int64()))
+	}
+	keccak.lock.Unlock()
+	if threads == 0 {
+		threads = runtime.NumCPU()
+	}
+	if threads < 0 {
+		threads = 0 // Allows disabling local mining without extra logic around local/remote
+	}
+	// Push new work to remote sealer
+	if keccak.workCh != nil {
+		keccak.workCh <- &sealTask{block: block, results: results}
+	}
+	var (
+		pend   sync.WaitGroup
+		locals = make(chan *types.Block)
+	)
+	for i := 0; i < threads; i++ {
+		pend.Add(1)
+		go func(id int, nonce uint64) {
+			defer pend.Done()
+			keccak.mine(block, id, nonce, abort, locals)
+		}(i, uint64(keccak.rand.Int63()))
+	}
+	// Wait until sealing is terminated or a nonce is found
+	go func() {
+		var result *types.Block
+		select {
+		case <-stop:
+			// Outside abort, stop all miner threads
+			close(abort)
+		case result = <-locals:
+			// One of the threads found a block, abort all others
+			select {
+			case results <- result:
+			default:
+				log.Warn("Sealing result is not read by miner", "mode", "local", "sealhash", keccak.SealHash(block.Header()))
+			}
+			close(abort)
+		case <-keccak.update:
+			// Thread count was changed on user request, restart
+			close(abort)
+			if err := keccak.Seal(chain, block, results, stop); err != nil {
+				log.Error("Failed to restart sealing after update", "err", err)
+			}
+		}
+		// Wait for all miners to terminate and return the block
+		pend.Wait()
+	}()
+	return nil
+}
+
+// mine is the actual proof-of-work miner that searches for a nonce starting from
+// seed that results in correct final block difficulty.
+func (keccak *Keccak) mine(block *types.Block, id int, seed uint64, abort chan struct{}, found chan *types.Block) {
+	// Extract some data from the header
+	var (
+		header = block.Header()
+		hash   = keccak.SealHash(header).Bytes()
+		target = new(big.Int).Div(two256, header.Difficulty)
+	)
+	// Start generating random nonces until we abort or find a good one
+	var (
+		attempts = int64(0)
+		nonce    = seed
+	)
+	logger := log.New("miner", id)
+	logger.Trace("Started keccak search for new nonces", "seed", seed)
+search:
+	for {
+		select {
+		case <-abort:
+			// Mining terminated, update stats and abort
+			logger.Trace("Keccak nonce search aborted", "attempts", nonce-seed)
+			keccak.hashrate.Mark(attempts)
+			break search
+
+		default:
+			// We don't have to update hash rate on every nonce, so update after after 2^X nonces
+			attempts++
+			if (attempts % (1 << 15)) == 0 {
+				keccak.hashrate.Mark(attempts)
+				attempts = 0
+			}
+			// TODO THIS IS WHERE POW IS
+			// Compute the PoW value of this nonce
+			// digest, result := hashimotoFull(dataset.dataset, hash, nonce)
+			sha := sha3.NewLegacyKeccak256().Write([]byte(hash))
+			result := sha.Sum(nil)
+			log.Info("XXXxxxxxxxxxxxxxxxxxxxMINING")
+
+			// if the result is bigger than the target
+			if new(big.Int).SetBytes(result).Cmp(target) == 1 {
```

Comment:
you probably want to test for `result < target`. the ethash code has `if new(big.Int).SetBytes(result).Cmp(target) <= 0 {` which is effectively checking for "less than" target. It looks like you have "greater than" target in this code

target is a huge uint256 that starts with a few zeros (because `target = new(big.Int).Div(two256, header.Difficulty)`) . the target is reached if your result hash starts with more zeros than it (i.e. result < target)